Query: Issue propert client eval warning for Last & LastOrDefault

smitpatel opened the pull request:
Resolves #6929 

Issue: we always set ClientEval as false even though it was required. Added proper handling.

Rules:
Last with order by can always be translated to Sql. Do not log warning.
LastOrDefault with order by gets translated but we still do client eval for the exception (just like FirstOrDefault), Do not log warning.
Warn when LastOrDefault is used inside subquery.
Warn when Last is used without order by because we need to evaluate on client to find the last record.

maumar commented on src/Microsoft.EntityFrameworkCore.Relational.Specification.Tests/WarningsTestBase.cs:
```diff
@@ -80,7 +80,62 @@ public virtual void Single_SingleOrDefault_without_orderby_doesnt_issue_warning(
-      
+
+        [Fact]
+        public virtual void Last_with_order_by_does_not_issue_client_eval_warning()
```
We should always client eval Last because in case collection is empty it should throw an exception. LastOrDefault we could translate safely

smitpatel replied:
What is current behavior with `First`?

smitpatel replied:
Do we issue client eval warning for top level `First()` (when it is not inside subquery)? Due to exception, we would evaluate it on client anyway.

divega replied:
`First()` in that case may be evaluated on the client to be able to throw but as long as the filtering happens on the server I believe it would be confusing to issue the warning.

maumar replied:
yeah I think you guys are right, my concern was that if someone builds on top of Last(), they would want to know that those additional operations are evaluated on the client, but we shall issue warning for those, when they come. Perhaps we could add this as a test, i.e. use Last in the subquery and then compose on top of the result

```

```

divega replied:
>  Perhaps we could add this as a test...

Sounds good to me.

maumar commented:
:shipit:

smitpatel commented:
@maumar @divega - I updated PR. There was small error which I corrected.
Current behavior is this:
If there is no order by then it will issue warning since we need to do client eval only.
If there is order by then
- `LastOrDefault` does not issue any warning.
- `Last` issues warning if its inside subquery. Top level `Last` does not issue warning.

I have updated test accordingly.
